packages/cosmic-swingset/lib/ag-solo/start.js

```diff
@@ -123,23 +123,42 @@ async function buildSwingset(
     }
   }
 
+  // This synchronises the processKernel calls between
+  // deliverInboundToMbx and deliverInboundCommand.
+  let lastProcessKernelP = Promise.resolve();
+  function processKernelP() {
+    // Enqueue a processKernel, and shorten the promise chain.
+    lastProcessKernelP = lastProcessKernelP.finally(
+      _ => (lastProcessKernelP = processKernel()),
+    );
+    return lastProcessKernelP;
+  }
+
   async function deliverInboundToMbx(sender, messages, ack) {
     if (!(messages instanceof Array)) {
       throw new Error(`inbound given non-Array: ${messages}`);
     }
     // console.log(`deliverInboundToMbx`, messages, ack);
     if (mb.deliverInbound(sender, messages, ack, true)) {
-      await processKernel();
+      await processKernelP();
     }
   }
 
   async function deliverInboundCommand(obj) {
     // this promise could take an arbitrarily long time to resolve, so don't
     // wait on it
+    let boxedException;
     const p = cm.inboundCommand(obj);
-    // TODO: synchronize this somehow, make sure it doesn't overlap with the
-    // processKernel() call in deliverInbound()
-    await processKernel();
+    p.catch(e => {
+      // Save the exception for us to reraise.
+      // We stuff it in an array in case of `throw undefined`.
+      boxedException = [e];
+    });
+    // Run our processKernel.
+    await processKernelP();
+    if (boxedException) {
```

Comment:
The interleaving here feels complicated and delicate.  In particular, there seems to be a presumption that if there's an exception in `cm.inboundCommand`, the `p.catch` closure will be run before the promise returned by `processKernelP` resolves.  I can well imagine that's true, but it's not _obviously_ true.  Contrariwise, if you're not expecting this then whether `boxedException` is undefined or not seems stochastic.  Or maybe I'm not understanding this at all.  Anyway, I think I'd like to see an explanatory comment.

Reply:
Good point.  I've simplified this and added some explanatory text.